server/market/market.go

```diff
@@ -3,30 +3,435 @@
+		case s := <-m.orderRouter:
+			if m.epochQueue == nil {
+				log.Debugf("Order %v received prior to market start.", s.rec.order)
+				s.errChan <- ErrMarketNotRunning
+				continue
+			}
+
+			// Ensure order's server time is in the current epoch, otherwise
+			// send it back and allow the epoch to cycle.
+			if s.sTime == nil {
+				now := time.Now().UTC()
+				// Retain the server time if it was sent back to the queue by a
+				// previous iteration of this loop.
+				s.sTime = &now
+			}
+
+			// If the order's server time stamp is not in the current epoch,
+			// send it back to the queue.
+			switch {
+			case m.epochQueue.IncludesTime(*s.sTime): // good, keep going
+				log.Debugf("Order for current epoch received: %v", s.rec.order)
+			case s.sTime.Before(m.epochQueue.Start): // order for epoch in the past - should not happen!
+				log.Errorf("Dropping order %v stamped for epoch in the past: %v!",
+					s.rec.order.ID(), s.sTime)
+				s.errChan <- ErrEpochMissed
+				continue
+			default: // for future epoch
+				log.Warnf("Order not in current epoch [%d,%d): %d",
+					m.epochQueue.Start.UnixNano(), m.epochQueue.End.UnixNano(), s.sTime.UnixNano())
+				// Order's server time falls in next epoch. Resubmit.
+				go func() {
+					time.Sleep(time.Until(nextEpoch.Start) + time.Millisecond)
+					m.orderRouter <- s
+				}() // cannot be blocking
+				continue
+			}
+
+			// Set the server time stamp here, BEFORE referencing the order ID.
+			ord := s.rec.order
+			ord.SetTime(s.sTime.Unix())
+
+			if ord := m.epochQueue.Orders[ord.ID()]; ord != nil {
+				// TODO: remove this check when we are sure the order router is behaving
+				log.Errorf("Received duplicate order %v!", ord)
+				s.errChan <- ErrDuplicateOrder
+				continue
+			}
+
+			epochIdx := uint64(m.epochQueue.Epoch)
+
+			// Validate the order. The order router must do it's own validation,
+			// but do a second validation for (1) this Market and (2) epoch
+			// status, before putting it on the queue.
+			if !m.validateOrder(ord) {
+				log.Errorf("(*Market).runEpochs: Invalid order received for epoch %d: %v",
+					epochIdx, ord)
+				s.errChan <- ErrInvalidOrder
+				continue
+			}
+
+			// Sign the order and prepare the client response. Only after the
+			// archiver has successfully stored the new epoch order should the
+			// order be committed for processing.
+			respMsg, err := m.orderResponse(s.rec, epochIdx)
+			if err != nil {
+				log.Errorf("failed to create msgjson.Message for order %v, msgID %v response: %v",
+					s.rec.order, s.rec.msgID, err)
+				s.errChan <- ErrMalformedOrderResponse
+				continue
+			}
+
+			// Archive the new epoch order BEFORE inserting it into the epoch
+			// queue, initiating the swap, and notifying book subscribers.
+			if err := m.storage.NewEpochOrder(ord); err != nil {
+				log.Errorf("(*Market).runEpochs: Failed to store new epoch order %v: %v",
+					ord, err)
+				s.errChan <- ErrInternalServer
+				m.cancel()
+				return
+			}
```

Comment:
This should be made async (e.g. sending to a storage queue goroutine).  I'm just uneasy about inserting the order into the queue and sending the sign response unless we were able to record it in the database since downstream operations are likely to fail if it isn't stored.  A possible solution is to execute all of the remaining loop into a goroutine or queue so that the epoch cycle may continue without delay.